### PR TITLE
Configurable event hub checkpointing

### PR DESCRIPTION
**Pull request by mamaso**
related to #1060

### CODE REVIEW

**Comment by paulbatum** [comment]
Do you have a script PR that would let me set the batches per checkpoint via host.json?

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs [hunk 1]
```diff
@@ -69,6 +69,11 @@ public EventHubConfiguration()
         }
 
         /// <summary>
+        /// Sets the frequency at which to checkpoint event batches.
+        /// </summary>
+        public int BatchesPerCheckpoint { get; set; } = 1;
```
Comment (tohling) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: Would it be more self-explanatory to name is BatchCountPerCheckpoint?

Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: Naming is totally up in the air, I struggled to come up with something I liked :)

**Comment by mamaso** [comment]
@paulbatum at present no. It also might make sense to expose this at the trigger attribute level (aka function level) instead of host level config @MikeStall probably has some input on that front.

I'd probably prefer the function granularity, but current event hub listener settings are all host level.

**Comment by paulbatum** [comment]
@mamaso Yeah I can see the argument for being this configurable per function. It feels a bit weird though, because all of our performance knobs are currently host level.

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs [hunk 2]
```diff
@@ -55,16 +57,46 @@ IEventProcessor IEventProcessorFactory.CreateEventProcessor(PartitionContext con
...
+            }
+        }
+
         // We get a new instance each time Start() is called. 
         // We'll get a listener per partition - so they can potentialy run in parallel even on a single machine.
         private class Listener : IEventProcessor
         {
             private readonly EventHubListener _parent;
             private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+            private Func<PartitionContext, Task> _checkpoint;
```
Comment (fabiocav) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs]: Make this `readonly`

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs (hunk 1 above)
Comment (fabiocav) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: I also like `CheckpointFrequency`.... `BatchesCheckpointFrequency` perhaps?

**Comment by mamaso** [comment]
@mathewc or @MikeStall can you take a look at this?

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs [hunk 3]
```diff
@@ -69,6 +69,11 @@ public EventHubConfiguration()
         }
 
         /// <summary>
+        /// Sets the frequency at which to checkpoint event batches.
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: >/// Sets the frequency at which to checkpoint event batches. [](start = 8, length = 60)

Can you expand on the comment to say what exactly this is counting?

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs [hunk 4]
```diff
@@ -55,16 +57,46 @@ IEventProcessor IEventProcessorFactory.CreateEventProcessor(PartitionContext con
             return new Listener(this);
         }
 
+        internal static Func<Func<Task>, Task> CreateCheckpointStrategy(int batchesPerCheckpoint)
+        {
+            switch (batchesPerCheckpoint)
+            {
+                // no checkpointing
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs]: > // no checkpointing [](start = 15, length = 20)

Why would we allow this?

**Comment by MikeStall** [comment]
With the recent BYOB extensibility work, the extension's initialize method gets the host.json and it should be able to handle this with no Script PR changes. 
[https://github.com/Azure/azure-webjobs-sdk-script/issues/1409](https://github.com/Azure/azure-webjobs-sdk-script/issues/1409)  is tracking this.l

---
In reply to: [293703567](https://github.com/Azure/azure-webjobs-sdk/pull/1106#issuecomment-293703567) [](ancestors = 293703567)

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs (hunk 4 above)
Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs]: You're right, it shouldn't be allowed - need to move the cursor somehow.

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/IEventHubProvider.cs [hunk 5]
```diff
@@ -8,6 +8,9 @@ namespace Microsoft.Azure.WebJobs.ServiceBus
     // Expose to binders / attributes so they can get the EventHub connections. 
     internal interface IEventHubProvider
     {
+        // Gets the checkpointing frequency
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/IEventHubProvider.cs]: Are you rebased on dev? I don't even see this IEventHubProvider type existing anymore

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubTriggerAttributeBindingProvider.cs [hunk 6]
```diff
@@ -62,7 +62,7 @@ public Task<ITriggerBinding> TryCreateAsync(TriggerBindingProviderContext contex
             Func<ListenerFactoryContext, bool, Task<IListener>> createListener =
              (factoryContext, singleDispatch) =>
              {
-                 IListener listener = new EventHubListener(factoryContext.Executor, eventHostListener, options, singleDispatch);
+                 IListener listener = new EventHubListener(factoryContext.Executor, eventHostListener, options, singleDispatch, _eventHubConfig.BatchCheckpointFrequency);
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubTriggerAttributeBindingProvider.cs]: Why not just pass in the entire config object to the listener, so we don't have to modify this param list each time we add a new knob?

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs (hunk 3 above)
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: nit: "Gets or sets..."

Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: perhaps: "Gets or sets the number of batches to process before checkpointing"

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs [hunk 7]
```diff
@@ -69,6 +69,11 @@ public EventHubConfiguration()
...
+        /// </summary>
+        public int BatchCheckpointFrequency { get; set; } = 1;
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: This setting applies to single dispatch mode as well?

Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: Yep - single dispatch will pull a batch -> spawn batchSize # of tasks -> checkpoint after all batches complete

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubTriggerAttributeBindingProvider.cs (hunk 6 above)
Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubTriggerAttributeBindingProvider.cs]: Reasonable - we source eventHostListener, options, and the checkpoint frequency from the config. A lot of this is testability, but as long as we're only pulling poco properties off the config it should be no big deal.

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/IEventHubProvider.cs (hunk 5 above)
Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/IEventHubProvider.cs]: haven't rebased, must have changed since then.

**Pull request merged**